### PR TITLE
Anton broilovskiy

### PR DESCRIPTION
**Pull request by roman-kh**

### CODE REVIEW

**Comment by roman-kh** [comment]
Не называй коммиты "Fix errors", "Error сorrection" - такие названия совсем не проясняют, какие изменения привносит данный коммит

**Comment by roman-kh** [comment]
5 коммитов с изменениями только в виде пробелов - это чересчур.
Изучи спецификацию yaml - http://www.yaml.org/spec/1.2/spec.html

File: task_01/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+"""Init"""
+import numpy as np
```
Comment (roman-kh) [inline task_01/__init__.py]: зачем?

File: task_01/mybatch.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
+"""MyBatch"""
```
Comment (roman-kh) [inline task_01/mybatch.py]: делай подробные и понятные doctsring'и

File: task_01/mybatch.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+import sys
+sys.path.append('..')
+import numpy as np
+import tensorflow as tf
+from dataset import Batch, action, model
```
Comment (roman-kh) [inline task_01/mybatch.py]: Соблюдай PEP8 - https://www.python.org/dev/peps/pep-0008/#imports

File: task_01/mybatch.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+def generate_linear_data(lenght=10):
```
Comment (roman-kh) [inline task_01/mybatch.py]: во-первых, length
во-вторых, во всей методах numpy, которые ты используешь этот параметр называется `size` - старайся соблюдать преемственность

File: task_01/mybatch.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    """
+    Generation of data for fit linear regression.
+
+    lenght - lenght of data.
+
+    return:
+    x - array [0..lenght]
```
Comment (roman-kh) [inline task_01/mybatch.py]: почему только 1 мерный? обычно данные многомерные

File: task_01/mybatch.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    y - array [0..lenght] with some random noize
+    """
+    y = np.linspace(0, 10, lenght)
```
Comment (roman-kh) [inline task_01/mybatch.py]: зачем только равномерно распределенные на отрезке?
генери случайные `X`, случайную ошибку `e`, определяй или тоже генери веса и формируй `Y` по формуле (`Y = Xw + e`)

File: task_01/mybatch.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    Generation of data for fit logistic regression.
+
+    lenght - lenght of data.
+
+    return:
+        x - random numbers from the range of -10 to 10
+    y - array of 1 or 0. if x[i] < 0 y[i] = 0 else y[i] = 1
+    """
+    x = np.array(np.random.randint(-10, 10, lenght), dtype=np.float32)
+    y = np.array([1. if i > 0 else 0. for i in x])
```
Comment (roman-kh) [inline task_01/mybatch.py]: это точно самый быстрый способ генерации `Y`?

File: task_01/mybatch.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+    x - matrix with shape(lenght,3) with random numbers of uniform distribution
+    """
+    x = np.random.random(lenght * 3).reshape(lenght, 3)
+    y = np.random.poisson(np.exp(np.dot(x, lam)))
+
+    return x, y
+
+class MyBatch(Batch):
+    """
+    Main class
```
Comment (roman-kh) [inline task_01/mybatch.py]: doctring'и должны быть понятными и полезными

File: task_01/mybatch.py [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+    """
+    def __init__(self, index, *args, **kwargs):
+        """
+        Initialization of variable from parent class - Batch.
+        """
+        super().__init__(index, *args, **kwargs)
+        self.x = None
+        self.y = None
+        self.w = None
+        self.b = None
```
Comment (roman-kh) [inline task_01/mybatch.py]: компоненты можно не инициализировать None'ами

File: task_01/mybatch.py [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+        self.loss = None
+
+    @property
+    def components(self):
+        """
+        Define componentis.
+        """
+        return 'x', 'y', 'w', 'b', 'loss'
```
Comment (roman-kh) [inline task_01/mybatch.py]: Батч состоит из элементов, каждый из которых состоит из компонентов
Компоненты можно рассматривать как столбцы матрицы батча
Поэтому `x`, `y` - хорошие примеры компонент, а `w`, `b` и `loss` - это вообще не компоненты и даже не атрибуты батча - это параметры модели

File: task_01/mybatch.py [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+
+    @action
+    def generate(self, lenght=10, ttype='linear', lam=np.array([0, 0, 0])):
```
Comment (roman-kh) [inline task_01/mybatch.py]: функцию генерации случайных данных лучше вынести из батча, поскольку к батчу она не имеет отношения

File: task_01/mybatch.py [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+
+        return: self
+        """
+        self = self.load(lenght=lenght, ttype=ttype, lam=lam)
+        self.x, self.y = self.x[self.indices], self.y[self.indices]
+
+        return self
+
+    @action
+    def load(self, lenght=10, ttype='linear', lam=np.array([0, 0, 0])):
```
Comment (roman-kh) [inline task_01/mybatch.py]: не меняй порядок и тип параметров переопределяемых методов, и уж тем более не меняй его характер.
метод `load` вызывается много где, и твой метод `load` сломает поведение батча

File: task_01/mybatch.py [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+        ttype - name of using algorithm:
+            * 'linear' for linear regression. (default)
+            * 'logistic' for logistic regression.
+            * 'poisson' for poisson regression.
+
+        return: self
+        """
+        data_dict = {'linear': generate_linear_data(lenght),
+                     'logistic': generate_logistic_data(lenght),
+                     'poisson': generate_poisson_data(lam, lenght)}
```
Comment (roman-kh) [inline task_01/mybatch.py]: в словарь вставляй не сгенерированные данные - это медленно и занимает лишнюю память, а ссылки на функции, которые генерят данные

File: task_01/mybatch.py [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+
+        return self
+
+    @model()
+    def linear_model():
+        """
+        Function with graph of linear regression.
+
+        return:
+        array with shape = (3,2)
```
Comment (roman-kh) [inline task_01/mybatch.py]: стремись создавать универсальные модели, не хардкодь размерность

File: task_01/mybatch.py [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+        w - slope coefficient of straight line.
+        b - bias.
+        """
+        x = tf.placeholder(name='input', dtype=tf.float32)
+        y = tf.placeholder(name='true_y', dtype=tf.float32)
+
+        w = tf.Variable(np.random.randint(-1, 1, size=1), name='weight', dtype=tf.float32)
+        b = tf.Variable(np.random.randint(-1, 1), dtype=tf.float32)
+
+        predict = tf.multiply(w, x, name='output') + b
```
Comment (roman-kh) [inline task_01/mybatch.py]: как это еще можно реализовать с помощью tensorflow?
